url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XExec.java

```diff
@@ -0,0 +1,1927 @@
+				callFrame.instructionpointer += i1 * 5;
+				i1 = callFrame.readUShort();
+				callFrame.instructionpointer += i1 * 4;
+				int type = callFrame.readUByte();
+				if (type >= 0 && type < 4) {
+					long v = v1.getInt();
+					long min = type == 0 ? callFrame.readByte() : type == 1 ? callFrame.readShort() : type == 2 ? callFrame.readIntP() : callFrame.readLongP();
+					int size = callFrame.readUShort();
+					v -= min;
+					if (v < 0 || v > size) {
+						callFrame.jumpTo(def);
+					} else {
+						callFrame.instructionpointer += v * 2;
+						i1 = callFrame.readUShort();
+						callFrame.jumpTo(i1);
+					}
+				} else if (type == 4) {
+					long v = v1.getInt();
+					int size = callFrame.readUShort();
+					int left = 0;
+					int right = size - 1;
+					int base = callFrame.instructionpointer;
+					while (left <= right) {
+						int m = (left + right) / 2;
+						callFrame.instructionpointer = base + m * 5;
+						int t = callFrame.readUByte();
+						long cv;
+						if (t == 0) {
+							cv = callFrame.readUShort();
+						} else if (t == 1) {
+							cv = callFrame.readIntP();
+						} else if (t == 2) {
+							cv = callFrame.readLongP();
+						} else {
+							throw new AssertionError();
+						}
+						if (cv == v) {
+							i1 = callFrame.readUShort();
+							callFrame.jumpTo(i1);
+							return;
+						} else if (cv < v) {
+							left = m + 1;
+						} else {
+							right = m - 1;
+						}
+					}
+					callFrame.jumpTo(def);
+				} else {
+					callFrame.jumpTo(def);
+				}
+			} else if (v1.isFloat()) {
+				int def = callFrame.readUShort();
+				callFrame.instructionpointer += 6;
+				int size = callFrame.readUShort();
+				double v = v1.getFloat();
+				int left = 0;
+				int right = size - 1;
+				int base = callFrame.instructionpointer;
+				while (left <= right) {
+					int m = (left + right) / 2;
+					callFrame.instructionpointer = base + m * 5;
+					int t = callFrame.readUByte();
+					double cv;
+					if (t == 0) {
+						cv = callFrame.readFloatP();
+					} else if (t == 1) {
+						cv = callFrame.readDoubleP();
+					} else {
+						throw new AssertionError();
+					}
+					if (cv == v) {
+						i1 = callFrame.readUShort();
+						callFrame.jumpTo(i1);
+						return;
+					} else if (cv < v) {
+						left = m + 1;
+					} else {
+						right = m - 1;
+					}
+				}
+				callFrame.jumpTo(def);
+			} else if (XUtils.isInstanceOf(rt, v1, rt.getBaseType(XUtils.STRING))) {
+				s = XUtils.getString(rt, v1);
+				int def = callFrame.readUShort();
+				callFrame.instructionpointer += 6;
+				i1 = callFrame.readUShort();
+				callFrame.instructionpointer += i1 * 5;
+				int size = callFrame.readUShort();
+				int left = 0;
+				int right = size - 1;
+				int base = callFrame.instructionpointer;
+				while (left <= right) {
+					int m = (left + right) / 2;
+					callFrame.instructionpointer = base + m * 4;
+					String cs = callFrame.readStringP();
+					int cmp = s == cs ? 0 : s.compareTo(cs);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.